### ISSUE TITLE
spack style broken due to py-isort bootstrap failure

### ISSUE THREAD

**Issue by luke-dt** [LLNL/spack#31265]
### Steps to reproduce

```console
$ spack clean -ab
$ spack style
```

### Error message

``` console
==> Running style checks on spack
  selected: isort, mypy, flake8
==> Modified files
==> Bootstrapping py-isort from pre-built binaries
==> Bootstrapping py-isort@4.3.5: %gcc target=x86_64 from sources
[+] /usr (external python-3.10-mbxrq3rxi6s4pf276aa6qyb3gtualou3)
==> Installing py-pip-21.3.1-j3v37odwspurlangep3l25ns2bnnwxz7
==> No binary for py-pip-21.3.1-j3v37odwspurlangep3l25ns2bnnwxz7 found: installing from source
==> Fetching https://mirror.spack.io/_source-cache/archive/de/deaf32dcd9ab821e359cd8330786bcd077604b5c5730c0b096eda46f95c24a2d
==> No patches needed for py-pip
==> py-pip: Executing phase: 'install'
==> py-pip: Successfully installed py-pip-21.3.1-j3v37odwspurlangep3l25ns2bnnwxz7
  Fetch: 0.34s.  Build: 1.26s.  Total: 1.60s.
[+] /home/lukedt/.spack/bootstrap/store/linux-fedora36-x86_64/gcc-12.1.1/py-pip-21.3.1-j3v37odwspurlangep3l25ns2bnnwxz7
==> Installing py-wheel-0.37.0-ara22q42psz6lhfll45vugzjyhsinl7d
==> No binary for py-wheel-0.37.0-ara22q42psz6lhfll45vugzjyhsinl7d found: installing from source
==> Fetching https://mirror.spack.io/_source-cache/archive/21/21014b2bd93c6d0034b6ba5d35e4eb284340e09d63c59aef6fc14b0f346146fd
==> No patches needed for py-wheel
==> py-wheel: Executing phase: 'install'
==> py-wheel: Successfully installed py-wheel-0.37.0-ara22q42psz6lhfll45vugzjyhsinl7d
  Fetch: 0.34s.  Build: 0.26s.  Total: 0.59s.
[+] /home/lukedt/.spack/bootstrap/store/linux-fedora36-x86_64/gcc-12.1.1/py-wheel-0.37.0-ara22q42psz6lhfll45vugzjyhsinl7d
==> Installing py-poetry-core-1.0.8-5fefs2b4ikojstbnrcaylvjkx64y55up
==> No binary for py-poetry-core-1.0.8-5fefs2b4ikojstbnrcaylvjkx64y55up found: installing from source
==> Fetching https://mirror.spack.io/_source-cache/archive/95/951fc7c1f8d710a94cb49019ee3742125039fc659675912ea614ac2aa405b118.tar.gz
==> No patches needed for py-poetry-core
==> py-poetry-core: Executing phase: 'install'
==> py-poetry-core: Successfully installed py-poetry-core-1.0.8-5fefs2b4ikojstbnrcaylvjkx64y55up
  Fetch: 0.37s.  Build: 0.95s.  Total: 1.33s.
[+] /home/lukedt/.spack/bootstrap/store/linux-fedora36-x86_64/gcc-12.1.1/py-poetry-core-1.0.8-5fefs2b4ikojstbnrcaylvjkx64y55up
==> Installing py-isort-5.9.3-k2i4if2s27nwicednjvpsejds5utaqmg
==> No binary for py-isort-5.9.3-k2i4if2s27nwicednjvpsejds5utaqmg found: installing from source
==> Fetching https://mirror.spack.io/_source-cache/archive/9c/9c2ea1e62d871267b78307fe511c0838ba0da28698c5732d54e2790bf3ba9899.tar.gz
==> No patches needed for py-isort
==> py-isort: Executing phase: 'install'
==> py-isort: Successfully installed py-isort-5.9.3-k2i4if2s27nwicednjvpsejds5utaqmg
  Fetch: 0.25s.  Build: 0.67s.  Total: 0.92s.
[+] /home/lukedt/.spack/bootstrap/store/linux-fedora36-x86_64/gcc-12.1.1/py-isort-5.9.3-k2i4if2s27nwicednjvpsejds5utaqmg
==> Error: cannot bootstrap any of the isort executables from spec "py-isort@4.3.5: %gcc target=x86_64" due to the following failures:
github-actions-v0.2 raised FileNotFoundError: [Errno 2] No such file or directory: '/home/lukedt/spack-dev/share/spack/bootstrap/github-actions-v0.2/py-isort.json'
github-actions-v0.1 raised ValueError: source is not trusted
Run `spack --debug ...` for more detailed errors
```

### Information on your system

* **Spack:** 0.19.0.dev0 (c64298a1835d51389ad5ed46a2713ba266a5c93b)
* **Python:** 3.10.5
* **Platform:** linux-fedora36-zen3
* **Concretizer:** clingo

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have searched the issues of this repo and believe this is not a duplicate
- [x] I have run the failing commands in debug mode and reported the output

**Comment by luke-dt** [comment]
**More detailed error message:**

```console
[+] /home/lukedt/.spack/bootstrap/store/linux-fedora36-x86_64/gcc-12.1.1/py-isort-5.9.3-k2i4if2s27nwicednjvpsejds5utaqmg
==> [2022-06-23-08:12:00.507885] Flagging py-isort-5.9.3-k2i4if2s27nwicednjvpsejds5utaqmg as installed
==> [2022-06-23-08:12:00.508752] Creating stage lock spack-stage-py-isort-5.9.3-k2i4if2s27nwicednjvpsejds5utaqmg
==> [2022-06-23-08:12:00.508855] Downgrading to a read lock on py-isort-5.9.3-k2i4if2s27nwicednjvpsejds5utaqmg with timeout 1e-09
==> [2022-06-23-08:12:00.508959] py-isort-5.9.3-k2i4if2s27nwicednjvpsejds5utaqmg is now read locked
==> [2022-06-23-08:12:00.509027] Releasing read lock on py-pip-21.3.1-j3v37odwspurlangep3l25ns2bnnwxz7
==> [2022-06-23-08:12:00.509096] Releasing read lock on py-wheel-0.37.0-ara22q42psz6lhfll45vugzjyhsinl7d
==> [2022-06-23-08:12:00.509159] Releasing read lock on py-poetry-core-1.0.8-5fefs2b4ikojstbnrcaylvjkx64y55up
==> [2022-06-23-08:12:00.509248] Releasing read lock on py-isort-5.9.3-k2i4if2s27nwicednjvpsejds5utaqmg
==> [2022-06-23-08:12:00.510962] [BOOTSTRAP EXECUTABLES isort] Try installed specs with query 'py-isort@5.9.3%gcc@12.1.1~colors arch=linux-fedora36-x86_64 ^py-pip@21.3.1%gcc@12.1.1 arch=linux-fedora36-x86_64 ^py-poetry-core@1.0.8%gcc@12.1.1 arch=linux-fedora36-x86_64 ^py-wheel@0.37.0%gcc@12.1.1 arch=linux-fedora36-x86_64 ^python@3.10%gcc@12.1.1+bz2+ctypes+dbm~debug+ensurepip+libxml2+lzma~nis~optimizations+pic+pyexpat+pythoncmd+readline+shared+sqlite3+ssl~tix~tkinter~ucs4+uuid+zlib patches=0d98e93,7d40923,f2fd060 arch=linux-fedora36-x86_64'
==> [2022-06-23-08:12:00.514295] Reading config file /home/lukedt/spack-dev/etc/spack/defaults/config.yaml
Traceback (most recent call last):
  File "/home/lukedt/./spack-dev/bin/spack", line 98, in <module>
    sys.exit(spack.main.main())
  File "/home/lukedt/spack-dev/lib/spack/spack/main.py", line 893, in main
    return _main(argv)
  File "/home/lukedt/spack-dev/lib/spack/spack/main.py", line 848, in _main
    return finish_parse_and_run(parser, cmd_name, env_format_error)
  File "/home/lukedt/spack-dev/lib/spack/spack/main.py", line 876, in finish_parse_and_run
    return _invoke_command(command, parser, args, unknown)
  File "/home/lukedt/spack-dev/lib/spack/spack/main.py", line 533, in _invoke_command
    return_val = command(parser, args)
  File "/home/lukedt/spack-dev/lib/spack/spack/cmd/style.py", line 422, in style
    commands[tool_name] = bootstrap_fn()
  File "/home/lukedt/spack-dev/lib/spack/spack/bootstrap.py", line 834, in ensure_isort_in_path_or_raise
    return ensure_executables_in_path_or_raise([executable], abstract_spec=root_spec)
  File "/home/lukedt/spack-dev/lib/spack/spack/bootstrap.py", line 611, in ensure_executables_in_path_or_raise
    raise RuntimeError(msg)
RuntimeError: cannot bootstrap any of the isort executables from spec "py-isort@4.3.5: %gcc target=x86_64" due to the following failures:
github-actions-v0.2 raised FileNotFoundError: [Errno 2] No such file or directory: '/home/lukedt/spack-dev/share/spack/bootstrap/github-actions-v0.2/py-isort.json'
  File "/home/lukedt/spack-dev/lib/spack/spack/bootstrap.py", line 587, in ensure_executables_in_path_or_raise
    if b.try_search_path(executables, abstract_spec):
  File "/home/lukedt/spack-dev/lib/spack/spack/bootstrap.py", line 378, in try_search_path
    data = self._read_metadata(abstract_spec.name)
  File "/home/lukedt/spack-dev/lib/spack/spack/bootstrap.py", line 283, in _read_metadata
    with open(json_path) as f:

github-actions-v0.1 raised ValueError: source is not trusted
  File "/home/lukedt/spack-dev/lib/spack/spack/bootstrap.py", line 584, in ensure_executables_in_path_or_raise
    source_is_enabled_or_raise(current_config)
  File "/home/lukedt/spack-dev/lib/spack/spack/bootstrap.py", line 472, in source_is_enabled_or_raise
    raise ValueError('source is not trusted')
```

**Comment by alalazo** [comment]
Here's a system where it works (Ubuntu 20.04):
```console
$ ls /home/culpo/.spack/bootstrap/store/linux-ubuntu20.04-x86_64/gcc-9.4.0/py-isort-5.9.3-5ycfo3cc5sjmb3occ4quqvfymjcvm564
bin  lib
```
and here's Fedora where this fails:
```console
[root@166c7158c565 spack]# ls /root/.spack/bootstrap/store/linux-fedora36-x86_64/gcc-12.1.1/py-isort-5.9.3-k2i4if2s27nwicednjvpsejds5utaqmg/
local
[root@166c7158c565 spack]# ls /root/.spack/bootstrap/store/linux-fedora36-x86_64/gcc-12.1.1/py-isort-5.9.3-k2i4if2s27nwicednjvpsejds5utaqmg/local
bin  lib  lib64
```
So, again, a different directory structure for Python installations. :cry:

**Comment by alalazo** [comment]
We have to make this line:

https://github.com/spack/spack/blob/c533612ab6e4d4b4c46d3b827fae3687302cd22a/lib/spack/spack/bootstrap.py#L207

more resilient to changes in the directory where executables are placed, since they might change across platforms. 

I can confirm that hacking that line and using:
```python
bin_dir = concrete_spec.prefix.local.bin 
```
makes things work in Fedora.

**Comment by manuelakuhn** [comment]
I get the same error for `spack style` but for me the reason is `py-black` instead of `py-isort` and `github-actions-v0.3` instead of `v0.2` (also on fedora36)

```console
[+] $HOME/.spack/bootstrap/store/linux-fedora36-x86_64/gcc-12.2.1/py-black-21.12b0-jj4xqyne4fvoctelkeg267f7ptiwpolp
==> Error: cannot bootstrap any of the black executables from spec "py-black@:21 %gcc target=x86_64" due to the following failures:
github-actions-v0.3 raised FileNotFoundError: [Errno 2] No such file or directory: '$SPACK/share/spack/bootstrap/github-actions-v0.3/py-black.json'
github-actions-v0.2 raised ValueError: source is not trusted
github-actions-v0.1 raised ValueError: source is not trusted
Run `spack --debug ...` for more detailed errors
```

Unfortunately the hack from @alalazo does not work for me. It results in the error:
```
==> Error: cannot bootstrap any of the black executables from spec "py-black@:21 %gcc target=x86_64" due to the following failures:
github-actions-v0.3 raised NoHeadersError: Unable to locate python headers in any of these locations:
    /usr/include/python3.10
/usr/include/3.10
/usr/Headers
github-actions-v0.2 raised ValueError: source is not trusted
github-actions-v0.1 raised ValueError: source is not trusted
spack-install raised NoHeadersError: Unable to locate python headers in any of these locations:
    /usr/include/python3.10
/usr/include/3.10
/usr/Headers
Run `spack --debug ...` for more detailed errors
```

Info on system:
* **Spack:** 0.19.0.dev0 (ddd18351a172f0aa4da1da202295eb7243f9c721)
* **Python:** 3.10.6
* **Platform:** linux-fedora36-skylake
* **Concretizer:** clingo

**Comment by luke-dt** [comment]
Confirming that I am also getting a similar error (except with isort). The suggested hack used to work but no longer works. System specs:

* **Spack:** 0.19.0.dev0 (ef33fdb4b4c7b7d261205f299bef1a564a96e192)
* **Python:** 3.10.6
* **Platform:** linux-fedora36-zen3
* **Concretizer:** clingo

**Comment by alalazo** [comment]
This should be solved by #34029
